packages/vite/src/node/optimizer/scan.ts

```diff
@@ -264,37 +262,40 @@ function esbuildScanPlugin(
               const src = srcMatch[1] || srcMatch[2] || srcMatch[3]
               js += `import ${JSON.stringify(src)}\n`
             } else if (content.trim()) {
-              // There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
+              // The reason why virtual modules are needed:
+              // 1. There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
               // or local scripts (`<script>` in Svelte and `<script setup>` in Vue)
+              // 2. There can be multiple module scripts in html
               // We need to handle these separately in case variable names are reused between them
-              const contextMatch = openTag.match(contextRE)
-              const context =
-                contextMatch &&
-                (contextMatch[1] || contextMatch[2] || contextMatch[3])
-              if (
-                (path.endsWith('.vue') && setupRE.test(openTag)) ||
-                (path.endsWith('.svelte') && context !== 'module')
-              ) {
-                // append imports in TS to prevent esbuild from removing them
-                // since they may be used in the template
-                const localContent =
-                  content +
-                  (loader.startsWith('ts') ? extractImportPaths(content) : '')
-                localScripts[path] = {
-                  loader,
-                  contents: localContent
+
+              // append imports in TS to prevent esbuild from removing them
+              // since they may be used in the template
+              const contents =
+                content +
+                (loader.startsWith('ts') ? extractImportPaths(content) : '')
+
+              const key = `${path}?id=${scriptId++}`
```

Comment:
I think we should have an incrementing `scriptId` for per-file basis. So every file's script would start with 1 first as it's a bit more predictable.

Reply:
I noticed you used a `scriptIdMap` to map the id. I think we can directly have a new incrementing number variable somewhere beside this spot:

https://github.com/vitejs/vite/blob/c226e860ae5fb2092ac042b8abe1858887057646/packages/vite/src/node/optimizer/scan.ts#L235

and use that number to create the `key` instead. We don't need to worry about the key clashing here as for every `path` esbuild loads, it is only processed once, so the `key` is always unique.

Reply:
Oh, right. I'm overthinking it. I'll change it later

Reply:
@bluwy  I have made some changes based on your newly comments, sorry for changing it several times 😂